agent/http_test.go

```diff
@@ -714,6 +714,42 @@ func TestEnableWebUI(t *testing.T) {
 	}
 }
 
+func TestHttpOptions(t *testing.T) {
+
+	req, _ := http.NewRequest("OPTIONS", "/v1/query", nil)
+	resp := httptest.NewRecorder()
+	a := NewTestAgent(t.Name(), "")
+	a.srv.Handler.ServeHTTP(resp, req)
+
+	// Check the result
+	if resp.Code != 200 {
+		t.Fatalf("should handle options method")
+	}
+	optionsStr := resp.Header().Get("Allow")
+	if optionsStr == "" {
+		t.Fatalf("options method should set 'Allow' header")
+	} else if optionsStr != "OPTIONS,GET,POST" {
+		t.Fatalf("options method should set 'Allow' header correctly")
+	}
+}
```

Comment:
Could you change this and `TestMethodNotAllowed` to loop through all the endpoints in allowedMethods and check their response codes/headers (instead of just the query/kv ones)?

Reply:
I realised from your comments that existing tests existed in `http_oss_test.go` so I moved my testing there.

This exposed an issue with specific prepared queries, since OPTIONS should return different methods if the path ends `/execute` or `/explain`. I got around that by allowing endpoints to register with no methods if they want to handle `OPTIONS`/`NotFoundError`s themselves ([this commit](https://github.com/hashicorp/consul/pull/3885/commits/c5f0bb371138aa534950db8b554a39b70813e0de)). Very open to other solutions if this isn't the cleanest way.